### PR TITLE
fix(augmentation): include padding in crop transforms

### LINKED ISSUE #4542: RandomCrop explicit padding still triggers oversized matrix scaling
RandomCrop rescales its transformation matrix against the unpadded input when the requested size exceeds either input axis and `pad_if_needed=False`. Explicit padding can already make the crop fit, but the correction still runs. Slice-mode image coordinates then disagree with keypoints/boxes; resample mode also distorts the image.

Reproduced on PR #4538 at `b9891a44`, Linux CPU, PyTorch 2.14.0:

```python
import torch
import kornia.augmentation as K
from kornia.geometry.keypoints import Keypoints

image = torch.arange(9.).reshape(1, 1, 3, 3)
crop = K.RandomCrop((4, 4), padding=1, p=1, cropping_mode="slice")
params = crop.forward_parameters(image.shape)
params["src"] = torch.tensor([[[0., 0.], [3., 0.], [3., 3.], [0., 3.]]])
output = crop(image, params=params)
points = Keypoints(torch.tensor([[[2., 2.]]]))
print(crop.transform_matrix)
print(crop.transform_keypoints(points, params, crop.flags, crop.transform_matrix).data)
```

The matrix is `diag(4/3, 4/3, 1)` and the keypoint becomes `(4, 4)`, outside the 4x4 output. The image pixel remains at `(3, 3)` after pre-padding. The crop matrix should be identity for these explicit source/destination corners; pre-padding already supplies the coordinate shift. With `cropping_mode="resample"`, the same incorrect scaling changes the bottom-right ramp value from 8 to 5.

Related: #4414 covers oversized crops without padding. PR #4538 will document and pin this separate explicit-padding case without changing runtime behavior.

Written by Codex (gpt-6-astra) on behalf of @ducha-aiki

### PR DESCRIPTION
**Pull request by dexhunter**
## What does this pull request do?

A 4×4 `RandomCrop` with `padding=1` on a 3×3 image applies a 4/3 scale correction even though the padded canvas is large enough. Slice-mode pixels then disagree with transformed keypoints and boxes; resample mode also distorts the image.

Use the effective padded canvas, including stored padding during parameter replay. Crops that fit after padding retain their sampled translation. Padded crops that remain oversized also change: requesting 10×10 from a 3×3 image with `padding=1` uses scale 10/5 instead of 10/3. The no-padding behavior tracked in #4414 is unchanged. Tensor division avoids an extra rounding step in float16 scales.

This review repair replaces the ineffective `extra_args` tests with actual padding-configuration changes during replay, adds coverage for parameters without `padding_size`, and uses literal expected pixels for the partially oversized case. It also corrects the docstring and changelog; the production algorithm is unchanged by this repair.

## Related issue or discussion

Fixes #4542.

## What did you check?

The final explicit-padding tests passed 32 cases across CPU float16, bfloat16, float32 and float64. The float16 pixel literals pin the observed rounded output of the previously published implementation; matrix expectations are specified separately as rational scales. The pre-precision-repair implementation still fails the float16 partial-fit test. Ruff, full cached pre-commit and type checking passed.

The new stored-padding and static-fallback tests cover both crop modes. Forcing the static branch makes the replay tests fail; deleting the static padding adjustment makes the fallback tests fail. The original implementation was restored after each mutation.

The original probe's four padding cases failed on the base and passed with the fix, with both controls passing. The later float16 partial-fit regression failed before the precision repair and passed afterward. An earlier four-dtype run of crop conventions and `RandomCrop` produced 151 passed, 8 skipped, 1 xfailed and 1 failed. Its remaining `test_pad_if_needed_both[cpu-bfloat16]` failure was identical before and after that repair.

[Original baseline/candidate evidence](https://dashboard.weco.ai/share/lGVyN6peWXtX9eWfxkNgXlDCcV1WI61g) · [Earlier replay repair](https://dashboard.weco.ai/share/p9wz9SMtffQ6Xp6yWMJcBE8NDKNWWQEs) · [Float16 repair](https://dashboard.weco.ai/share/-y8KEMAMbt3u94R6eZh1lIo5jdSA387B)

Local checks used Linux CPU. Accelerator behavior was not exercised locally. The pre-existing fullgraph compile limitation remains outside this repair.

## How did AI help?

GPT-5.6 Luna prepared the implementation and tests. GPT-6 Astra reviewed the source and validation evidence, rejected the ignored flag overrides as coverage, and requested literal pixel expectations and clearer documentation of the padded oversized case.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Review of `0d505df7`.

The fix is right and I verified it broadly. Reproduced #4542 on main with a round-trip probe — `RandomCrop((4,4), padding=1)` on a 3x3 `arange(9)/8`: keypoint `(2,2)` goes to `(4.0, 4.0)`, outside the 4x4 output, while the pixel is at `(3,3)`; matrix diag `1.3333333730697632` — and confirmed the head puts it at `(3,3)` with an identity matrix in both cropping modes. A 3200-config round-trip sweep (every `padding` form × `padding_mode` × `fill` × `pad_if_needed` × 5 sizes × 2 modes × 5 seeds, all nine pixels pushed through the returned matrix) takes the "padded canvas fits but round trip broken" bucket from **450 to 0**, leaves the genuinely-oversized bucket unchanged at 480 (#4414), and the `inverse()` acid test for `size=(6,6), padding=1` improves from `0.268` to `8.94e-07`. Reverting only `crop.py` fails all 16 new/changed nodes, and the batch-6b wart pin was correctly **renamed and inverted** rather than left contradicting the new ones (`4/3`→`1`, `[[4,4]]`→`[[3,3]]`) with no assertion weakened or dropped. Merged with current main (`30dfbf711`, post-#4566) the merge is clean, `pytest -k 4542` is 12 passed, full `tests/augmentation` is 3299 passed with the same 5 pre-existing XPASSes as main, MPS float32 passes, and neither half-precision manifest needs an edit. Five of seven mutations are killed.

Two blocking asks, one assertion each. To be clear up front: I am not disputing the implementation, only that two of its three branches can be deleted without any test noticing.

### [P2] Pin the tensor / stored-canvas branch

Replacing `kornia/augmentation/_2d/geometric/crop.py:208` with `if True:` deletes that whole branch and leaves the *entire* suite green:

```
MUT[always-static-path] full tests/augmentation -> 5534 passed, 6776 skipped, 42 xfailed, 7 xpassed
```

while reproducing #4542 exactly. With params generated at `padding=1` and the padding configuration then changed, the mutant returns `matrix diag 1.3333333730697632` against a canvas that was still padded; the head returns `1.0`.

The test meant to cover this, `test_random_crop_explicit_padding_replay_uses_stored_canvas_under_flag_overrides`, cannot: `extra_args` arrive as plain `**kwargs` (`kornia/augmentation/container/ops.py:300`) and the forward swallows unknown keywords, so `{"padding": 0, "pad_if_needed": False}` overrides nothing.

```
kwargs ignored? out equal: True   matrix equal: True
self.flags['padding'] after kwarg call: 1
bogus kwarg: ACCEPTED AND IGNORED        # totally_bogus_kwarg=42
```

As written that test is a duplicate of `test_random_crop_explicit_padding_preserves_matrix_4542` with extra ceremony, and the same applies to the `extra_args={"pad_if_needed": True}` half of `test_random_crop_explicit_padding_partial_fit_uses_padded_canvas_4542`. Either pin the real divergence (generate params, change the padding configuration, then replay), or — if no supported API can make the stored canvas differ from the recomputed one — drop the branch and keep a single path.

### [P2] Pin the static/export branch

Deleting `crop.py:211-212` (`h += padding[2] + padding[3]` / `w += padding[0] + padding[1]`) also leaves the whole suite green (`5534 passed`), while both reachable entries regress: a `params` dict without `"padding_size"` returns the buggy `1.3333333730697632` matrix, and `torch.export` output stops matching eager bitwise (`bitwise out: False`, versus `bitwise=True maxdiff=0` on the head). A params-without-`padding_size` case is a three-line test.

### [P3] Expand the changelog fragment

Beyond #4542's repro, the correction in the *still*-oversized-after-padding region changes silently:

```
RandomCrop((10,10), padding=1) on 3x3: matrix diag 3.333333 -> 2.000000
RandomCrop((6,4),  padding=1) on 3x3: matrix diag (1.333333, 2.000000) -> (0.800000, 1.200000)
```

Measured over 2304 configurations, 576 change output — all of them with `pad_if_needed=False`, explicit non-zero `padding`, and a size exceeding the unpadded input; nothing with `padding=None`/`0` or `pad_if_needed=True` moves a single bit. `.fixed` is the right type given the configuration was already self-inconsistent and pinned as a wart (the same shape as `changelog.d/4495.fixed.md`, not `3945.breaking.md`'s broad default-output change), but the fragment should state old and new behavior explicitly, including the `10/3 -> 10/5` scale change.

### [P3] Docstring nits

At `crop.py:92-96`, `#4414` is now linked twice in the same paragraph with a stray line break (`"remains\nunchanged."`) — docutils reports `Duplicate explicit target name: "#4414"` at INFO level, so Sphinx `-W` will not fail on it. And "The no-padding oversized behavior tracked in #4414 remains unchanged" is incomplete: the with-padding oversized behavior *does* change, to the padded-canvas denominator. Worth getting right, because `docs.yml` is path-filtered to `docs/**` so no `-W` build checked this prose.

### [P3] Minor

`test_random_crop_explicit_padding_partial_fit_uses_padded_canvas_4542` builds its expected pixels by calling `resize` / `crop_by_transform_mat` — the production primitives — instead of pinning a literal array like the rest of the file. It does pin the matrix entries (`4/5`, `6/5`) as literals separately, so the risk is contained, but a literal expected array would be a stronger pin.

For the record: `torch.compile(fullgraph=True)` fails identically on main and head (`GuardOnDataDependentSymNode` from the pre-existing `padding_size.unique(dim=0).cpu()...tolist()`), so no compile regression; `grep -rnE "#4542|issues/4542" kornia/ tests/ docs/ README.md` is clean on the merged tree (main had one hit at `crop.py:95`); `crop.py` defines only `RandomCrop`, with `CenterCrop`/`RandomResizedCrop` in separate files taking no padding; and `ruff@0.16.7`, the doctest, `check_changelog.py` and `closingIssuesReferences = [#4542]` are all correct.


Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Review by ducha-aiki — APPROVED** [review]
## Re-review — #4588, head `fdfe6f6b` — **approving**

**Both blocking P2s are discharged**, and each of the two branches that survived the whole 5534-node suite last round is now killed by a dedicated test. Merge onto `ecbcad561` is clean (`0f5a1e632`).

### The two mutations that survived last round

| mutation | last round | now |
| --- | --- | --- |
| `crop.py:208` → `if True:` (force the static path, deleting the stored-canvas branch) | survived all 5534 tests | **killed** — 6 failed, all `test_random_crop_explicit_padding_replay_uses_stored_canvas_under_flag_overrides[…]`, greatest abs diff `0.333` |
| delete `crop.py:211-212` (`h += padding[2]+[3]` / `w += padding[0]+[1]`) | survived | **killed** — 6 failed, all `test_random_crop_explicit_padding_recomputes_static_canvas_without_saved_padding[…]` |

### The `extra_args` problem — resolved by removal, and I re-confirmed why that was the right move

You dropped `extra_args` from both tests and replaced it with `crop.flags["padding"] = 0` for the stored-canvas branch and `params.pop("padding_size")` for the static one. The underlying wart is still there, so avoiding it was correct: an `extra_args={DataKey.INPUT: {"padding": 0, "pad_if_needed": False}}` override leaves the output **identical** with `crop.flags["padding"]` still `1`, and a `totally_bogus_kwarg=42` is accepted and silently ignored both as a direct forward kwarg and via `extra_args`.

I also checked the deeper path, since "silently swallowed" could have hidden a divergence: `override_parameters` *does* write kwargs into the local `flags`, but for `RandomCrop` the stored `params["padding_size"]` dominates in **both** `apply_transform` and `compute_transformation`, so `crop(x, padding=0)` is inert *consistently* — image and matrix stay in agreement. No divergence introduced.

### The blast radius, independently recounted

2304-config A/B sweep (6 `padding` forms × 2 `pad_if_needed` × 6 sizes × 2 modes × 2 input shapes × 8 seeds), hashing output **and** `transform_matrix`: **448 changed**, all with `pad_if_needed=False` and explicit non-zero padding. Split:

- **176** are the #4542 repro — fits after padding, now identity
- **272** are still-oversized-after-padding — the denominator changes

Nothing with `padding=None`, `padding=0`, or `pad_if_needed=True` moves a bit (1856 configs bit-identical).

Independent round-trip probe over 186 fitting configs, pushing 9 pixel coordinates through the returned matrix and comparing against the image: **broken 66 → 0**.

`torch.export.export` on `RandomCrop` with padding, 3 configs: exports and matches eager bitwise on both trees; the matrix diagonal goes `1.3333333730697632 → 1.0` for `(4,4)/padding=1`, and `(1.3333333, 2.0) → (0.8, 1.2)` for `(6,4)/padding=1`.

---

## Three P3s — none blocking

**The five new tests are the only non-prefixed tests in any convention file.** They are `test_random_crop_*`, neither `test_convention_*` nor `test_wart_*`. Across the five `tests/augmentation/test_conventions*.py` files, `geometric_crop` has `total=19, non-prefixed=5`; the other four have `non-prefixed=0` out of 153. Rename to `test_convention_random_crop_*`, keeping the `_4542` suffixes.

**The fragment now describes only the oversized-scale change, not the defect #4542 actually reported.** The old text said "keeping image and coordinate transforms consistent"; the new one says only "uses the effective pre-crop canvas for oversized requests … `10/3` to `10/5`". That covers the 272 still-oversized configs but not the 176 in which image and matrix-transformed keypoints/boxes silently disagreed — the 66 → 0 round-trip result above. One clause:

> …and keypoints, boxes and masks now land on the same pixels as the cropped image when the crop fits after padding (#4542).

**The float16 `resample` expected table is captured implementation output, not a derived value** (`test_..._partial_fit_uses_padded_canvas_4542`, :221-231), with a comment naming CPU/PyTorch 2.9.1. I measured the generic table's deviation as `max|out-generic| = 1.587e-3` in float16 — over the float16 tolerance, hence your branch — and `7.81e-3` in bfloat16, under it, hence no branch. The captured table matches exactly on torch 2.14.0. It is self-referential, and the blocking half-precision leg runs 2.9.1, which I cannot exercise here. Risk is contained (the matrix is pinned independently and the slice-mode table is dtype-independent), but asserting the float16 case with an explicit loosened `atol` against the generic table would be sturdier than a snapshot.

**One note, not an ask:** the `is_exporting()` disjunct at `crop.py:208` is not load-bearing on torch 2.14 — removing it still exports all three padded configs, still matches eager bitwise, and the crop file stays at 85 passed. Unlike the two branches above, that is an *equivalent* mutant here, i.e. defensive rather than untested. I did not measure it under dynamic shapes or ONNX/dynamo export. Recording it so a future reader does not mistake the survivor for a gap.

---

## Also verified

**6c debt still done correctly on the new head.** `test_wart_random_crop_explicit_padding_scales_matrix_4542` → `test_random_crop_explicit_padding_preserves_matrix_4542`, same fixture, every assertion **inverted rather than weakened**: matrix `4/3 → 1`, points `[[4,4]] → [[3,3]]`, boxes `(boxes+1)*(4/3) → boxes+1`, and the two mode-specific expected arrays collapsing to one because resample no longer distorts. `test_wart_random_crop_oversized_rescales_both_matrix_axes_4414` is untouched and passes — correct, since #4414 is out of scope and the new code still rescales both axes when either is oversized. `grep -rnE "#4542|issues/4542" kornia/ tests/ docs/` on the merge tree: **no hits** (base had exactly one, `crop.py:95`).

**Last round's two smaller P3s are fixed**: the duplicate `#4414` explicit link target is gone, the "remains unchanged" sentence now carries the explicit `10/3 → 10/5` statement, and the partial-fit test no longer builds its expectation from `resize`/`crop_by_transform_mat` — both imports dropped.

**Asymmetric padding + oversize is correct.** `RandomCrop((8,8), padding=(2,1))` on 3×3 gives `scale_w = 1.142857 = 8/7` and `scale_h = 1.600000 = 8/5` on both the stored and the static branch; index order (`padding_size[:, 2:4]` → H, `[:, 0:2]` → W) verified.

**Counts.** Crop conventions file at float32/float64/float16: 85 passed, 0 failed; bfloat16: 33 passed; MPS float32: 26 passed. Related surface (`container/test_random_crop.py`, `test_lazy_matrix.py`, `test_torch_export.py`, `TestRandomCrop`) at 3 dtypes: 306 passed. Full `tests/augmentation --dtype=float32`: **3340 passed** vs base 3332, identical xfail/xpass (+8 = the 4 new/renamed nodes × 2 `mode` params). Broad `-k "crop or Crop or pad or Pad"` across `tests/augmentation` + `tests/geometry/transform/test_crop2d.py`: PR 4 failed / 810 passed, base 4 failed / 786 passed — **the same four** pre-existing cpu-float16 nodes.

**torch 2.5.1 overlay**: float32+float64 → 0 failures on the PR. The float16 deltas there are all `grid_sampler_2d_cpu not implemented for Half`, and the half-precision CI job pins 2.9.1 while the 2.5.1 matrix legs run only float32/float64 (`pr_test_cpu.yml:89-151`), so that never reaches CI.

No `testing/half_precision_xfails/` or `known_failure_xfails/mps_float32.txt` entry is owed — none mentions this file. Pinned `ruff@0.16.7` clean, doctest passes, fragment correctly named, nothing over 120.

## Mutation testing

| mutation | result |
| --- | --- |
| `crop.py:208` → `if True:` | **killed** — 6 failed |
| delete `crop.py:211-212` | **killed** — 6 failed |
| `needs_scale` `\|` → `&` | **killed** — 18 failed |
| swap `transform[:,0,0] *= scale_w` / `[:,1,1] *= scale_h` | **killed** — 24 failed |
| delete both `transform[:,i,i] *= scale_*` | **killed** — 24 failed |
| swap H/W padding components in *both* branches | survives the crop file (85 passed) but **killed at directory level** — `tests/augmentation --dtype=float32` goes 2 failed (`TestRandomCrop::test_pad_if_needed_width`/`_height`). Non-equivalent: returns `scale_w=1.600000, scale_h=1.142857`, exactly transposed. A `padding=(2,1)` + oversized case in the new file would localize it. |
| `>` → `>=` in `needs_scale` | survived — **equivalent**: at equality the scale is exactly `1.0` |
| drop the `is_exporting() or` guard | survived — equivalent on the measured configs, see above |

Posted on behalf of @ducha-aiki by Claude (Opus 5).
